tsdb/shard.go

```diff
@@ -1786,136 +1827,192 @@ func (fs *MeasurementFieldSet) IsEmpty() bool {
 		}
 		if b == nil {
 			// No fields, file removed, all done
 			return true, nil
 		}
 		if _, err := fd.Write(b); err != nil {
-			return true, err
+			return true, fmt.Errorf("failed saving fields to %s: %w", path, err)
 		}
 		return false, fd.Sync()
 	}()
-	if err != nil || isEmpty {
-		return
+	if err != nil {
+		return err
+	} else if isEmpty {
+		// remove empty file
+		return os.RemoveAll(fs.path)
 	}
-	err = fs.renameFile(path)
+	return fs.renameFile(path)
 }
 
-// marshalMeasurementFieldSet: remove the fields.idx file if no fields
-// otherwise, copy the in-memory version into a protobuf to write to
-// disk
-func (fs *MeasurementFieldSet) marshalMeasurementFieldSet() ([]byte, error) {
-	fs.mu.Lock()
-	defer fs.mu.Unlock()
-	if len(fs.fields) == 0 {
-		// If no fields left, remove the fields index file
-		if err := os.RemoveAll(fs.path); err != nil {
-			return nil, err
-		} else {
-			return nil, nil
+// writeChangesToFile: Write a change file for fields.idx
+func (fscm *measurementFieldSetChangeMgr) writeChangesToFile(first writeRequest) {
+	var err error = nil
+	// Put the errorChannel on which we blocked into a slice to allow more invocations
+	// to share the return code from the file write
+	errorChannels := []errorChannel{first.errorReturn}
+	changes := []FieldChanges{first.changes}
+	// On return, send the error to every go proc that send changes
+	defer func() {
+		for _, c := range errorChannels {
+			c <- err
+			close(c)
+		}
+	}()
+	// Do some blocking IO operations before marshalling the changes,
+	// to allow other changes to be queued up and be captured in one
+	// write operation, in case we are under heavy field creation load
+	fscm.mu.Lock()
+	defer fscm.mu.Unlock()
+	fd, err := os.OpenFile(fscm.Path, os.O_CREATE|os.O_APPEND|os.O_SYNC|os.O_WRONLY, 0666)
+	if err != nil {
+		err = fmt.Errorf("failed opening %s: %w", fscm.Path, err)
+		return
+	}
+
+	// ensure file closed
+	defer errors2.Capture(&err, fd.Close)()
+
+	// Read all the pending field and measurement write or delete
+	// requests
+	for {
+		select {
+		case wr := <-fscm.writeRequests:
+			changes = append(changes, wr.changes)
+			errorChannels = append(errorChannels, wr.errorReturn)
+			continue
+		default:
 		}
+		break
+	}
+	// marshal the alice of slices of field changes
+	var b []byte
+	b, err = marshalFieldChanges(changes...)
+	if err != nil {
+		err = fmt.Errorf("failed marshaling changes to %s: %w", fd.Name(), err)
+		return
+	}
+	if b == nil {
+		// No fields, nothing to do but close the file.
+		return
+	}
+	if err = writeSizePlusBuffer(fd, b); err != nil {
+		err = fmt.Errorf("failed writing to %s: %w", fd.Name(), err)
+		return
+	}
+	if err = fd.Sync(); err != nil {
+		err = fmt.Errorf("failed syncing %s: %w", fd.Name(), err)
+	}
+}
+
+func writeSizePlusBuffer(w io.Writer, b []byte) error {
+	var numBuf [8]byte
+	binary.LittleEndian.PutUint64(numBuf[:], uint64(len(b)))
+	if _, err := w.Write(numBuf[:]); err != nil {
```

Comment:
If this fails the log file is corrupted right? What does the caller do with the error?

Reply:
The error goes back to the sender of the write request, like other write errors.

Reply:
What happens to the next write? Does it recover somehow, or is the file permanently damaged?

If the file is permanently damaged, I don't think just failing the write that caused the damage is sufficient. The TSL files do:

```
	// Write record to file.
	n, err := f.w.Write(f.buf)
	if err != nil {
		// Move position backwards over partial entry.
		// Log should be reopened if seeking cannot be completed.
		if n > 0 {
			f.w.Reset(f.file)
			if _, err := f.file.Seek(int64(-n), io.SeekCurrent); err != nil {
				f.Close()
			}
		}
		return err
	}
```

I don't think the `Close` is a good idea - at that point we should probably just abort the process, closing the log file in the middle of the append call hopefully just causes every subsequent write to fail... but the `Seek` is good.

Reply:
We don't hold the file handle open between writes, so I would call `File.Truncate` then `File.Close` to eliminate the bad write, if we want this strategy.

Reply:
I think it is important to maintain crash-safety for this on-disk structure, such that in theory the database can crash at any time without causing data loss / corruption

Reply:
Per our discussion yesterday, working on that design at the moment.